server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -55,13 +58,21 @@
 public class TransportClusterStatsAction extends TransportNodesAction<ClusterStatsRequest, ClusterStatsResponse,
         TransportClusterStatsAction.ClusterStatsNodeRequest, ClusterStatsNodeResponse> {
 
+    private static final Logger logger = LogManager.getLogger(TransportClusterStatsAction.class);
+
     private static final CommonStatsFlags SHARD_STATS_FLAGS = new CommonStatsFlags(CommonStatsFlags.Flag.Docs, CommonStatsFlags.Flag.Store,
         CommonStatsFlags.Flag.FieldData, CommonStatsFlags.Flag.QueryCache,
         CommonStatsFlags.Flag.Completion, CommonStatsFlags.Flag.Segments);
 
     private final NodeService nodeService;
     private final IndicesService indicesService;
 
+    // guards #mappingStats, #analysisStats and #metaVersion
+    private final Object statsMutex = new Object();
+
+    private MappingStats mappingStats;
+    private AnalysisStats analysisStats;
+    private long metaVersion = -1L;
```

Comment:
I'm absolutely out of practice here, so please do correct me if I'm wrong, but should `metaVersion` be `volatile`?

Reply:
No need for that. We only ever touch that field under the `statsMutex` so we are guaranteed to always have a consistent view of it.